src/heretic/utils.py

```diff
@@ -20,13 +22,114 @@
 from datasets.download.download_manager import DownloadMode
 from datasets.utils.info_utils import VerificationMode
 from optuna import Trial
+from questionary import Choice
 from rich.console import Console
 
 from .config import DatasetSpecification, Settings
 
 print = Console(highlight=False).print
 
 
+def is_notebook() -> bool:
+    # Check for specific environment variables (Colab, Kaggle)
+    # This is necessary because when running as a subprocess (e.g. !heretic),
+    # get_ipython() might not be available or might not reflect the notebook environment.
+    if os.getenv("COLAB_GPU") or os.getenv("KAGGLE_KERNEL_RUN_TYPE"):
+        return True
+
+    # Check IPython shell type (for library usage)
+    try:
+        from IPython import get_ipython
+
+        shell = get_ipython()
+        if shell is None:
+            return False
+
+        shell_name = shell.__class__.__name__
+        if shell_name in ["ZMQInteractiveShell", "Shell"]:
+            return True
+
+        if "google.colab" in str(shell.__class__):
+            return True
+
+        return False
+    except (ImportError, NameError, AttributeError):
+        return False
+
+
+def prompt_select(message: str, choices: list[Any], style=None) -> Any:
+    if is_notebook():
+        print()
+        print(message)
+        real_choices = []
+        for i, choice in enumerate(choices, 1):
+            if isinstance(choice, Choice):
+                print(f"[{i}] {choice.title}")
+                real_choices.append(choice.value)
+            else:
+                print(f"[{i}] {choice}")
+                real_choices.append(choice)
+
+        while True:
+            try:
+                selection = input("Enter number: ")
+                idx = int(selection) - 1
+                if 0 <= idx < len(real_choices):
+                    return real_choices[idx]
+                print(
+                    f"[red]Please enter a number between 1 and {len(real_choices)}[/]"
+                )
+            except ValueError:
+                print("[red]Invalid input. Please enter a number.[/]")
+    else:
+        return questionary.select(message, choices=choices, style=style).ask()
+
+
+def prompt_text(
+    message: str,
+    default: str = "",
+    password: bool = False,
+    unsafe: bool = False,
+    qmark: str = None,
+) -> str:
+    if is_notebook():
+        print()
+        if password:
+            return getpass.getpass(message)
```

Comment:
Please move password handling into `prompt_password`. There is no need to clutter this function with unrelated functionality.